File: ruby/api_requests.rb
require 'net/http'
require 'uri'
require 'benchmark'
require 'thread'

# Make a single HTTP request
def make_request(url, request_id)
  uri = URI.parse(url)
  response = Net::HTTP.get_response(uri)
  {
    id: request_id,
    status: response.code.to_i,
    success: response.code.to_i == 200
  }
rescue => e
  {
    id: request_id,
    status: 0,
    success: false,
    error: e.message
  }
end

# Run API request benchmark with 50 concurrent requests
def run_benchmark
  url = "https://jsonplaceholder.typicode.com/posts/1"
  num_requests = 50

  # Measure total time for concurrent requests
  start_time = Process.clock_gettime(Process::CLOCK_MONOTONIC)

  threads = []
  results = []
  mutex = Mutex.new

  num_requests.times do |i|
    threads << Thread.new do
      result = make_request(url, i)
      mutex.synchronize do
        results << result
      end
    end
  end

  threads.each(&:join)

  end_time = Process.clock_gettime(Process::CLOCK_MONOTONIC)
  execution_time = ((end_time - start_time) * 1000).round(2)

  # Count successful requests
  successful = results.count { |r| r[:success] }

  puts "Test: API Requests (50 concurrent)"
  puts "Total requests: #{num_requests}"
  puts "Successful requests: #{successful}"
  puts "Execution time: #{execution_time} ms"
end

if __FILE__ == $0
  run_benchmark
end

File: rust/src/api_requests.rs
use reqwest;
use std::time::Instant;
use tokio;

/// Make a single HTTP request
async fn make_request(url: &str, request_id: usize) -> (usize, bool) {
    match reqwest::get(url).await {
        Ok(response) => (request_id, response.status().is_success()),
        Err(_) => (request_id, false),
    }
}

/// Run API request benchmark with 50 concurrent requests
pub async fn run_benchmark() {
    let url = "https://jsonplaceholder.typicode.com/posts/1";
    let num_requests = 50;

    // Measure total time for concurrent requests
    let start_time = Instant::now();

    let mut handles = vec![];
    for i in 0..num_requests {
        let handle = tokio::spawn(make_request(url, i));
        handles.push(handle);
    }

    let mut successful = 0;
    for handle in handles {
        if let Ok((_, success)) = handle.await {
            if success {
                successful += 1;
            }
        }
    }

    let execution_time = start_time.elapsed().as_millis();

    println!("Test: API Requests (50 concurrent)");
    println!("Total requests: {}", num_requests);
    println!("Successful requests: {}", successful);
    println!("Execution time: {} ms", execution_time);
}
